# Section C (KNN Classification)

## k Nearest Neighbors (KNN)
### Introduction
K-Nearest Neighbors（KNN）is a supervised machine learning algorithm that can be used to solve both classification and regression problems.  It is also one of the best-known classification algorithms. The principle is that known data are arranged in a space defined by the selected features. When a new data is supplied to the algorithm, the algorithm will compare the classes of the k closest data to determine the class of the new data. It is based on the assumption that data points will fall into the vicinity of the same class of data. Simply put, similar things are close to each other. KNN is a lazy algorithm, because it doesn’t learn from the training dataset but “memorises” the training dataset instead, like putting labeled data in some metric space. KNN works by finding the distances between a query and all the examples in the data, selecting the specified number examples (K) closest to the query, then votes for the most frequent label (in the case of classification) or averages the labels (in the case of regression).


## Pros of K Nearest Neighbors
1. KNN is simple and intuive
KNN is a simple algorithm and hence easy to interpret the prediction.
2. KNN makes no assumptions about input data 
Non-parametric means there is no assumption for underlying data distribution. In other words, the model structure is determined from the dataset. This will be very helpful in practice where most of the real-world datasets do not follow mathematical theoretical assumptions. e.g., in Linear Regression we assume the linear dependency of labels on features.
it is a non-parametric model so makes no assumption about the underlying data pattern.Therefore, k-Nearest Neighbors is often the first option when there is a little or no prior knowledge about the data distribution.
3. KNN does not need the training step
The training phase of the algorithm consists only in storing the feature vectors and class labels of the training samples. This also means that the training stage is pretty fast, since data storage is all that is actually happening. The training step is much faster for the nearest neighbors compared to other machine learning algorithms.
4. KNN can be used for both classification and Regression.
The natural practice of this method is placing input points into appropriate categories, but we can easily extend this algorithm to regression problems. Instead of combining the discrete labels of k-neighbors we have to combine continuous predictions. These predictions can be obtained in different ways, for example, by averaging the values of the k-most similar instances.



## Cons of K Nearest Neighbors
1. KNN needs high memory.
Since it KNN memorise the input data instead of learn from them, it  needs high memory to store all the data points.

2. Prediction stage is costly.
However, when we predict new data using KNN, it will take much time to search for the nearest neighbors in the entire training dataset, calculating the distance and sorting 
them, so it is computationally expensive as it searches the nearest neighbors for the new point at the prediction stage and thus the prediction stage is very costly.

3. KNN is sensitive to outliers
Besides,k-NN classifier is sensitive to outliers, accuracy is impacted by noise or irrelevant data. Its accuracy highly depends on the quality of the training data. Noise and mislabeled data, as well as outliers and overlaps between data regions of different classes, lead to less accurate classification. it is  sensitive to outliers, since a single mislabeled example dramatically changes the class boundaries. Anomalies affect the method significantly, because k-NN gets all the information from the input, rather than from an algorithm that tries to generalize data. This problem can be dealt with by adopting either a large k value or by pre-processing the training set with an editing algorithm.  

4. KNN is not suitable for high dimensional data.
When features vector dimension is relatively high(a dataset in which the number of features p is larger than the number of observations N), k-NN might suffer. This is due to the fact that distance measures often lose accuracy: There is a little difference between the nearest and the farthest neighbors. Besides, irrelevant features and redundant features 
can affect its performance. we could use dimensionality reduction techniques (feature selection and feature extraction, like PCA) to redefine the features vector before using it as the input of the algorithm.

## Time and Space Complexity Analysis
n: number of points in the training dataset
d: data dimensionality
K: number of neighbours that we consider for voting
- Traning time complexity is O(1): The training phase of the algorithm consists only of storing the feature vectors and class labels of the training samples. Since  all computation is done during prediction, so KNN has O(1) time complexity.
- Traning space complexity is O(n*d): it needs to storage so the space occupied is O(n*d)
- Prediction time complexity:O(k*n*d): O(d) to compute sistance to one example, o(n*d) to find one nearest neighbor, O(k*n*d) to find k closet examples, thus the complexity is O(k*n*d)
- Prediction space complexity:O(1): Find the most frequent label in the list set just needs contant space


# improvements to reduce complexity of KNN
## dimensionality reduction
- feature selection
- pca
## use efficient method to find nearst neighbot
- In contrast with the fast training stage, it requires very expensive testing. All the cost of the algorithm is in the computation of the prediction, since for every test sample, the model has to run through the entire data set to compute distances and then find the nearest neighbors.

Proposal: To deal with the challenging computation of k-NN, some alternative techniques to the brute-force have been created. K-D tree and Ball tree are the most well-known ones.
## preprocessiong
normalized features
### How KNN works(Classification)
KNN algorithms can be summarised in the following steps:

1. Select the number k of the nearest neighbours
2. Calculate the distance between new data point and each of the training data.
3. Find the k nearest neighbours from the new data
4. Take the most frequent of labels as the prediction of the label of new data point

KNN algorithm can be used for both classification and regression problems. 

### How KNN predits
For a given new data point, find the k nearest neighbours from new data point in traning dataset, collect neighbours' labels, finally, take the mode of the labels as the prediction for the label of new data point.

### KNN Hyperparameters
Hyperparameters are the parameters to control the model learning process. We can tune these hyperparameters to get a model with optimal performance.
Here are part of hyperparmeters that I used in the following task.

- Number of neighbors(K):the number of nearest neighbors
- Weights: When calculating the distance, we can choose 'uniform’ weights that all points in each neighborhood are weighted equally or
‘distance’ weight points by the inverse of their distance.
- p: This determines that we can use what kind of metric to calculate distance between neighbours(Euclidean Distance, Manhattan Distance, and Minkowski Distance) When p = 1, use manhattan_distance, and euclidean_distance  p = 2. For arbitrary p, minkowski_distance is used.

### Application
KNN can be used in both solve both classification and regression problems, when using KNN to solve regression problem, it will use the average of target value of K nearst neighbors as the prediction value of the new data, when using to solve classification problem, it will use the mode as the prediction label of new data.


## Dataset
Here is the introduction of the dataset. 

This dataset is about 11 clinical features for predicting heart disease events, from Kaggle [Heart Failure Prediction Dataset](https://www.kaggle.com/fedesoriano/heart-failure-prediction/)


### Attribute Information:
- Age: (Discreate variable)    age of the patient [years]
- Sex: (Binary variable)   sex of the patient [M: Male, F: Female]
- ChestPainType: (Categorical variable) chest pain type [TA: Typical Angina, ATA: Atypical Angina, NAP: Non-Anginal Pain, ASY: Asymptomatic]
- RestingBP:(Continuous variable)  resting blood pressure [mm Hg]
- Cholesterol:(Continuous variable)    serum cholesterol [mm/dl]
- FastingBS: (Binary variable)     fasting blood sugar [1: if FastingBS > 120 mg/dl, 0: otherwise]
- RestingECG: (Categorical variable)   resting electrocardiogram results [Normal: Normal, ST: having ST-T wave abnormality, LVH: showing probable or definite left ventricular hypertrophy by Estes' criteria]
- MaxHR:(Continuos variable)   maximum heart rate achieved [Numeric value between 60 and 202]
- ExerciseAngina: (Binary variable)    exercise-induced angina [Y: Yes, N: No]
- Oldpeak: (Continuous variable)   Numeric value measured in depression
- ST_Slope: (Categorical variable)     he slope of the peak exercise ST segment [Up: upsloping, Flat: flat, Down: downsloping]
- HeartDisease:(Binary variable)   output class [1: heart disease, 0: Normal]

The classifaction goal is to predict heart falure.

In [1]:
import pandas as pd
df = pd.read_csv("heart.csv", sep = ',')

The dataset has 918 observations and 12 features (11 independent features and 1 dependent feature)

In [585]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


There are no missing values in these dataset.

In [586]:
print(df.isnull().sum())

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64


The target class distribution is relatively balanced.

In [16]:
df.iloc[:,-1].value_counts()/len(df.iloc[:,-1])

1    0.553377
0    0.446623
Name: HeartDisease, dtype: float64

Split the dataset into Feature vectors (X) and labels (y)

In [587]:
df.describe().T

count        mean         std   min     25%    50%    75%    max
Age           918.0   53.510893    9.432617  28.0   47.00   54.0   60.0   77.0
RestingBP     918.0  132.396514   18.514154   0.0  120.00  130.0  140.0  200.0
Cholesterol   918.0  198.799564  109.384145   0.0  173.25  223.0  267.0  603.0
FastingBS     918.0    0.233115    0.423046   0.0    0.00    0.0    0.0    1.0
MaxHR         918.0  136.809368   25.460334  60.0  120.00  138.0  156.0  202.0
Oldpeak       918.0    0.887364    1.066570  -2.6    0.00    0.6    1.5    6.2
HeartDisease  918.0    0.553377    0.497414   0.0    0.00    1.0    1.0    1.0

In [588]:
X = df.iloc[:,:-1]
y = df.iloc[:,-1]

## Convert categorical values
The dataset contains categorical variables like 'ChestPainType'

Since meachine learning algorithms cannot work with categorical data directly, we need to convert the categorical values into numeric values by using one-hot code.

To handle the categorical variables we have to create dummy variables out of a categorical variable and include them instead of the original categorical variable. Unlike regression, create k dummies instead of (k-1).

In [589]:
# Applying One hot Encoding on the Categorical columns
X=pd.get_dummies(X)
pd.DataFrame(X).head()

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  Sex_F  Sex_M  \
0   40        140          289          0    172      0.0      0      1   
1   49        160          180          0    156      1.0      1      0   
2   37        130          283          0     98      0.0      0      1   
3   48        138          214          0    108      1.5      1      0   
4   54        150          195          0    122      0.0      0      1   

   ChestPainType_ASY  ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  \
0                  0                  1                  0                 0   
1                  0                  0                  1                 0   
2                  0                  1                  0                 0   
3                  1                  0                  0                 0   
4                  0                  0                  1                 0   

   RestingECG_LVH  RestingECG_Normal  RestingECG_ST  ExerciseAngina_N  \
0               0                  1              0                 1   
1               0                  1              0                 1   
2               0                  0              1                 1   
3               0                  1              0                 0   
4               0                  1              0                 1   

   ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
0                 0              0              0            1  
1                 0              0              1            0  
2                 0              0              0            1  
3                 1              0              1            0  
4                 0              0              0            1

Since we don’t want the model to over-learn from training data and perform poorly after being tested in test data. we need to assess how well the model is generalizing. Hence, so we  separate input data set into training, validation, and testing subsets to prevent the model from overfitting and to evaluate the model effectively.

Train dataset is used to fit the parameters to the model, or 'finding nearest neighbors'

Valid dataset is used to assess the model's performance and avoid overfitting (the model is really good at classifying the samples in the training dataset but cannot generalize and make accurate classifications on the unseen data), it helps us tune the model's hyperparameters, so we can improve the model according to the results on validation.

test dataset is used to assess the performance of the final chosen model and it is equivalent to the future unseen data points.

## Split dataset 
we split dataset into train dataset, validation dataset, and test dataset in the ratio 8:1:1.

In [590]:
from sklearn.model_selection import train_test_split

# train dataset 80%, valid dataset 10%, test dataset 10%
# split the data in training and remaining dataset, and split the remaining dataset
X_remain, X_test, y_remain, y_test = train_test_split(X, y, test_size = 0.1, random_state = 3)

X_train, X_valid, y_train, y_valid = train_test_split(X_remain, y_remain, test_size = 1/9,random_state = 3)


## Feature Scaling

Scaling the data is important for kNN as bringing all the features to the same scale is recommended for applying distance-based algorithms. If the scale of features is very different then normalization is required. This is because the distance calculation done in KNN uses feature values. When the one feature values are large than other, that feature will dominate the distance hence the outcome of the KNN. In this dataset, the 'Age' attribute is from 28 to 77, the 'Cholesterol' is from 0 to 603. Since both the features have different scales, there is a chance that higher weightage is given to features with higher magnitude. This will impact the performance of the machine learning algorithm and obviously, we do not want our algorithm to be biassed towards one feature like 'Cholesterol' in this case.

Therefore, we scale our data before employing a distance based algorithm so that all the features contribute equally to the result, we need to scale our dataset.

I use Data Standardization method to resclae the attributes so that they have mean as 0 and variance as 1, this can bring down all the features to a common scale without distorting the differences in the range of the values.

I have imported standard scaler from scikit-learn machine learning library software. I call fit_transform() method on our training and validation dataset and transform() method on the test dataset.

## Metric

Choosing the right metric is crucial while evaluating machine learning models. 

Accuracy is an intuitive metric to assess the perfomance of model, and is defined as the number of correct predictions divided by the total number of predictions, multiplied by 100.

accuracy is applied in when the target class distribution is fairly balanced and we do not prefer any type of class. So I decide to use 'accuracy_score' function for this classication task.

## Train KNN model

I use 'KNeighborsClassifier' function form 'sklearn' library to generate KNN model.
First, I use train dataset to fit the KNN model and iterate the hyperparameters that we want to tuning (p, k, weigth) and in each for loop tries a combination of hyperparameters. The model tuning the hyperparameters based on it's perfomance (assessed by classification accuracy score) on validation data set, and finally we record the best hyperparameters for the training model.

### Hyperparameters tuning
p parameter for the distance metric. When p = 1, this is equivalent to using manhattan_distance, and euclidean_distance for p = 2. For arbitrary p (eg. p=3), minkowski_distance is used.

weights{‘uniform’, ‘distance’}
Weight function used in prediction. Possible values:
‘uniform’ : uniform weights. All points in each neighborhood are weighted equally.
‘distance’ : weight points by the inverse of their distance. in this case, closer neighbors of a query point will have a greater influence than neighbors which are further away.

In [591]:
from sklearn.neighbors import KNeighborsClassifier

best_p = 1
best_score = 0.0
best_k = 1
weight_option = ["uniform", "distance"]
k_range = list(range(3,31,2))
p_options = [1,2,3]

for p_o in p_options:
    for k in k_range:
        for weight in weight_option:
            knn = KNeighborsClassifier(n_neighbors = k, weights = weight, p = p_o)
            knn.fit(X_train, y_train)
            y_valid_pred = knn.predict(X_valid)
            score = knn.score(X_valid, y_valid) # Return the mean accuracy on the validate dataset
            if score > best_score:
                best_p = p_o
                best_k = k
                best_weight = weight
                best_score = score
                
print("best_p=", best_p)
print("best k=", best_k)
print("best weight=", best_weight)
print("best score=", best_score)
print('Test set accuracy: %.2f%%' % accuracy)


best_p= 1
best k= 25
best weight= uniform
best score= 0.7934782608695652


### Test KNN model on test dataset

I get 86.96% prediction accuracy on test dataset.

In [592]:
from sklearn.metrics import accuracy_score

knn = KNeighborsClassifier(n_neighbors = 25, p = 1, weights= 'uniform')

knn.fit(X_train, y_train)

y_test_pred=knn.predict(X_test) 

test_accuracy = accuracy_score(y_test,y_test_pred)

print('test_accuracy : %.4f'%test_accuracy)

test_accuracy : 0.8696


After use Standaization, The accuracy increases.

In [593]:
from sklearn.model_selection import train_test_split

# train dataset 80%, valid dataset 10%, test dataset 10%
# split the data in training and remaining dataset, and split the remaining dataset
X_remain, X_test, y_remain, y_test = train_test_split(X, y, test_size = 0.1, random_state = 3)

X_train, X_valid, y_train, y_valid = train_test_split(X_remain, y_remain, test_size = 1/9,random_state = 3)

from sklearn.preprocessing import StandardScaler

sc_x=StandardScaler()
X_train = sc_x.fit_transform(X_train)
X_valid = sc_x.fit_transform(X_valid)
X_test=sc_x.transform(X_test)

In [594]:
from sklearn.neighbors import KNeighborsClassifier

best_p = 1
best_score = 0.0
best_k = 1
weight_option = ["uniform", "distance"]
k_range = list(range(3,31,2))
p_options = [1,2,3]

for p_o in p_options:
    for k in k_range:
        for weight in weight_option:
            knn = KNeighborsClassifier(n_neighbors = k, weights = weight, p = p_o)
            knn.fit(X_train, y_train)
            y_valid_pred = knn.predict(X_valid)
            score = knn.score(X_valid, y_valid) # Return the mean accuracy on the validate dataset
            if score > best_score:
                best_p = p_o
                best_k = k
                best_weight = weight
                best_score = score
                
print("best_p=", best_p)
print("best k=", best_k)
print("best weight=", best_weight)
print("best score=", best_score)

best_p= 1
best k= 9
best weight= uniform
best score= 0.9021739130434783


In [595]:
from sklearn.metrics import accuracy_score

knn = KNeighborsClassifier(n_neighbors = 9, p = 1, weights= 'uniform')

knn.fit(X_train, y_train)

y_test_pred=knn.predict(X_test) 

test_accuracy = accuracy_score(y_test,y_test_pred)

print('test_accuracy : %.4f'%test_accuracy)

test_accuracy : 0.9130


## Cross-validation

Now, let's try use cross-validation to tune the hyperparameters.

There are a lot of different techniques that may be used to cross-validate a model. In this task, I choose K-folds validation to conduct cross-validation and set k = 5, here is the how it works.

Dataset is randomly split up into 'k' folds. One of the folds is validation set and the rest are training set. The model is trained on the training set and scored on the validation set to find best hyperparameters. Then the process is repeated until each group has been used as the validation set.

### Train KNN_2 model
I use 'GridSearchCV' function from 'sklearn' library to find the best set of hyperparameters that I tuning(k,p,distance).
There are four parameters I specify in this function: estimator ,param_grid,scoring,cv. I will explain how it use.

estimator:  the model that I train
param_grid: In the grid search method I specified with the 'param_grid‘ grid parameter with possible values for hyperparameters.when “fitting” it on a dataset all the possible combinations of parameter values are evaluated records the model performance (accuracy score). Finally, it returns the best KNN model with the best hyperparameters.I use accuracy score to choose the best hyperparameters combination.
Scoring: I choose accuracy score to access the model performance.  

CV: Strategy to evaluate the performance of the cross-validated model on the test set. I choose 5-fold cross validation as cross-validation splitting strategy. Here, the data set is split into 5 folds. In the first iteration, the first fold is used to test the model and the rest are used to train the model. In the second iteration, 2nd fold is used as the testing set while the rest serve as the training set. This process is repeated until each fold of the 5 folds have been used as the testing set.


In [596]:
from sklearn.model_selection import train_test_split

# train dataset 80%, valid dataset 10%, test dataset 10%
# split the data in training and remaining dataset, and split the remaining dataset
X_remain, X_test, y_remain, y_test = train_test_split(X, y, test_size = 0.1, random_state = 3)

X_train, X_valid, y_train, y_valid = train_test_split(X_remain, y_remain, test_size = 1/9,random_state = 3)

In [597]:
from sklearn.model_selection import GridSearchCV

# Hyperparameters that we want to tune.
k_range = list(range(3,31,2))
weight_options = ["uniform", "distance"]
p_options=[1,2,3]

param_dict = dict(n_neighbors = k_range, weights = weight_options, p = p_options)

#Create KNN classiifier
knn_2 = KNeighborsClassifier()

#Use 5 fold cross-validation, use accuracy to evaluate the performance of the model on the validation set
grid = GridSearchCV(estimator = knn_2, param_grid = param_dict, scoring = 'accuracy', cv = 5) 

#Fit the model (remain dataset includes train and validation dataset that I split before.)
best_model = grid.fit(X_remain,y_remain) 

#Print the value of best hyperparameters and best accuracy score
print('best hyperparameters', best_model.best_params_)
print('accuracy_score : %.4f' % best_model.best_score_ )

best hyperparameters {'n_neighbors': 17, 'p': 1, 'weights': 'distance'}
accuracy_score : 0.7748


In [602]:
from sklearn.metrics import accuracy_score

knn_2 = KNeighborsClassifier(n_neighbors = 17, p = 1, weights= 'distance')

knn_2.fit(X_remain, y_remain)

y_test_pred=knn_2.predict(X_test) 

test_accuracy = accuracy_score(y_test,y_test_pred)

print('test_accuracy : %.6f'%test_accuracy)

test_accuracy : 0.934783


After use standardization

In [603]:

from sklearn.preprocessing import StandardScaler

sc_x=StandardScaler()
X_remain = sc_x.fit_transform(X_remain)
X_test=sc_x.transform(X_test)

In [604]:
from sklearn.model_selection import GridSearchCV

# Hyperparameters that we want to tune.
k_range = list(range(3,31,2))
weight_options = ["distance","uniform"]
p_options=[1,2,3]

param_dict = dict(n_neighbors = k_range, weights = weight_options, p = p_options)

#Create KNN classiifier
knn_2 = KNeighborsClassifier()

#Use 5 fold cross-validation, use accuracy to evaluate the performance of the model on the validation set
grid = GridSearchCV(knn_2, param_grid = param_dict, scoring = 'accuracy', cv = 5) 

#Fit the model (remain dataset includes train and validation dataset that I split before.)
best_model = grid.fit(X_remain,y_remain) 

#Print the value of best hyperparameters and best accuracy score
print('best hyperparameters', best_model.best_params_)
print('accuracy_score : %.4f' % best_model.best_score_ )

best hyperparameters {'n_neighbors': 29, 'p': 1, 'weights': 'distance'}
accuracy_score : 0.8716


### Test KNN_2 model on test dataset

After tuning the hyperparameters by crossvalidation, I get another model with different hyperparameters, then I test the best performance model on test dataset and get 88.04% prediction accuracy, which is better than before(86.96%).

In [605]:
from sklearn.metrics import accuracy_score

knn_2 = KNeighborsClassifier(n_neighbors = 29, p = 1, weights= 'distance')

knn_2.fit(X_remain, y_remain)

y_test_pred=knn_2.predict(X_test) 

test_accuracy = accuracy_score(y_test,y_test_pred)

print('test_accuracy : %.6f'%test_accuracy)

test_accuracy : 0.945652


The second model that trained by cross validation has a higher accuracy score than the first model on the same test dataset. 
That is because when splitting the origin dataset into three sets, which reduce the amount of data in training date which can be used for learning the model, and the results can depend on a particular random choice for the pair of (training, validation) sets. But when useing cross validation, we can use of training data more effiently as every observation is used for both training and validating, thus get more accurate estimate of unseen data. 
However, it takes 13.8s to run cross-validation and it takes only 3.2 to run on one validation dataset, because K-fold cross-validation repeats the train/test split K-times, so it need more time to train the model and tune hyperparameters.
